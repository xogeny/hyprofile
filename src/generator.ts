import fs from "fs";
import path from "path";

export interface Generator {
    writeFile(filename: string, contents: string): Promise<void>;
}

export class FileWriter implements Generator {
    constructor(public directory: string) {}
    writeFile(filename: string, contents: string): Promise<void> {
        return new Promise<void>((resolve, reject) => {
            let file = path.join(this.directory, filename);
            fs.writeFile(file, contents, err => {
                console.log("File " + file + " written: ", err);
                if (err) reject(err);
                else resolve(undefined);
            });
        });
    }
}

export class MemoryWriter implements Generator {
    files: { [name: string]: string } = {};
    writer: Generator | undefined = undefined;
    constructor(directory?: string) {
        if (directory) {
            this.writer = new FileWriter(directory);
        }
    }
    async writeFile(filename: string, contents: string): Promise<void> {
        this.files[filename] = contents;
        if (this.writer) await this.writer.writeFile(filename, contents);
        return undefined;
    }
}
