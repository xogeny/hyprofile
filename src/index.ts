export * from "./profile";
export * from "./sample";
export * from "./elaborate";
export * from "./documentation";
export * from "./code";
