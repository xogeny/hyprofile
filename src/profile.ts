import { ISchema } from "@xogeny/ts-schema";

export type SchemaReference = { $ref: string };

export type Schema = ISchema | SchemaReference;

export interface Profile {
    resources?: { [className: string]: ClassProfile };
    relations?: { [rel: string]: Array<RelationProfile> };
    schemas?: { [schema: string]: Schema };
}

export interface ClassProfile {
    properties?: Schema;
    actions?: { [actionName: string]: ActionProfile };
}

export interface ActionProfile {
    title?: string;
    method?: "GET" | "POST" | "PUT" | "PATCH" | "OPTIONS" | "DELETE";
    type?: string;
    fields?: Schema;
}

export interface RelationProfile {
    from: string;
    to: string[];
}
