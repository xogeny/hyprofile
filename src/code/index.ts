import { ElaboratedProfile } from "../elaborate";
import { Generator } from "../generator";

export async function generateCode(profile: ElaboratedProfile, generator: Generator) {
    let contents = "";
    profile.resources.forEach(resource => {
        contents = `${contents}\n${resource.typedef}`;
    });
    await generator.writeFile("types.ts", contents);
}
