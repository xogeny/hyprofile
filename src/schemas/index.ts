import { Generator } from "../generator";
import { ElaboratedProfile } from "../elaborate";
import { JSONSchema4 } from "json-schema";

interface ResourceData {
    propSchema: JSONSchema4;
    actionSchemas: { [name: string]: JSONSchema4 };
}

interface SchemaData {
    resources: { [id: string]: ResourceData };
}

export async function generateSchemas(profile: ElaboratedProfile, generator: Generator): Promise<void> {
    let data: SchemaData = { resources: {} };
    profile.resources.forEach(resource => {
        data.resources[resource.resource] = {
            propSchema: resource.schema,
            actionSchemas: {},
        };
    });
    await generator.writeFile("schemas.json", JSON.stringify(data, null, 4));

    return undefined;
}
