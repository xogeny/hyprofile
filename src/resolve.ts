import { JSONSchema4 } from "json-schema";
import { dereference } from "json-schema-to-typescript/dist/src/resolver";
import { Schema, Profile } from "./profile";

export function resolveSchema(schema: Schema, profile: Profile): Promise<JSONSchema4> {
    const cwd = process.cwd() + "/";
    console.log("Current directory: ", cwd);
    if (typeof schema === "string") {
        let id = schema;
        if (profile.schemas) {
            schema = profile.schemas[id];
            return dereference(schema, cwd);
        } else {
            throw new Error("No local schema for " + id);
        }
    }
    return dereference(schema, cwd);
}
