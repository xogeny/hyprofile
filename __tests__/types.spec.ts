import * as hyprofile from "../src";

jest.setTimeout(200000);

describe("Sample", () => {
    it("should fail", async () => {
        let profile = hyprofile.Samples.sample1;
        let results = await hyprofile.elaborate(profile);
        console.log("results = ", results);
        expect(results).toMatchSnapshot();
    });
});
