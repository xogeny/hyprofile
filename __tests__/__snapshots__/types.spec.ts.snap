// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sample should fail 1`] = `
Object {
  "relations": Array [
    Object {
      "from": "order",
      "max": Infinity,
      "min": 0,
      "rel": "collection",
      "to": "order-items",
    },
    Object {
      "from": "order",
      "max": Infinity,
      "min": 0,
      "rel": "items",
      "to": "order-items",
    },
    Object {
      "from": "order",
      "max": Infinity,
      "min": 0,
      "rel": "customer",
      "to": "customer",
    },
    Object {
      "from": "order",
      "max": Infinity,
      "min": 0,
      "rel": "info",
      "to": "customer",
    },
    Object {
      "from": "order",
      "max": 1,
      "min": 0,
      "rel": "next",
      "to": "order",
    },
    Object {
      "from": "order",
      "max": 1,
      "min": 0,
      "rel": "prev",
      "to": "order",
    },
  ],
  "resources": Array [
    Object {
      "description": "An order submitted via the e-commerce API",
      "diagram": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"no\\"?><svg xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" contentScriptType=\\"application/ecmascript\\" contentStyleType=\\"text/css\\" height=\\"231px\\" preserveAspectRatio=\\"none\\" style=\\"width:361px;height:231px;\\" version=\\"1.1\\" viewBox=\\"0 0 361 231\\" width=\\"361px\\" zoomAndPan=\\"magnify\\"><defs><filter height=\\"300%\\" id=\\"fxhzbrp8gaet7\\" width=\\"300%\\" x=\\"-1\\" y=\\"-1\\"><feGaussianBlur result=\\"blurOut\\" stdDeviation=\\"2.0\\"/><feColorMatrix in=\\"blurOut\\" result=\\"blurOut2\\" type=\\"matrix\\" values=\\"0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 .4 0\\"/><feOffset dx=\\"4.0\\" dy=\\"4.0\\" in=\\"blurOut2\\" result=\\"blurOut3\\"/><feBlend in=\\"SourceGraphic\\" in2=\\"blurOut3\\" mode=\\"normal\\"/></filter></defs><g><!--class order--><rect fill=\\"#FEFECE\\" filter=\\"url(#fxhzbrp8gaet7)\\" height=\\"86.8652\\" id=\\"order\\" style=\\"stroke: #A80036; stroke-width: 1.5;\\" width=\\"136\\" x=\\"105\\" y=\\"8\\"/><ellipse cx=\\"152.75\\" cy=\\"24\\" fill=\\"#B4A7E5\\" rx=\\"11\\" ry=\\"11\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><path d=\\"M148.6777,19.7651 L148.6777,17.6069 L156.0571,17.6069 L156.0571,19.7651 L153.5918,19.7651 L153.5918,27.8418 L156.0571,27.8418 L156.0571,30 L148.6777,30 L148.6777,27.8418 L151.1431,27.8418 L151.1431,19.7651 Z \\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"12\\" font-style=\\"italic\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"32\\" x=\\"173.25\\" y=\\"28.5352\\">order</text><line style=\\"stroke: #A80036; stroke-width: 1.5;\\" x1=\\"106\\" x2=\\"240\\" y1=\\"40\\" y2=\\"40\\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"11\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"124\\" x=\\"111\\" y=\\"54.6348\\">orderNumber : number</text><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"11\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"107\\" x=\\"111\\" y=\\"67.5898\\">itemCount : number</text><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"11\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"74\\" x=\\"111\\" y=\\"80.5449\\">status : string</text><line style=\\"stroke: #A80036; stroke-width: 1.5;\\" x1=\\"106\\" x2=\\"240\\" y1=\\"86.8652\\" y2=\\"86.8652\\"/><!--class order-items--><rect fill=\\"#FEFECE\\" filter=\\"url(#fxhzbrp8gaet7)\\" height=\\"48\\" id=\\"order-items\\" style=\\"stroke: #A80036; stroke-width: 1.5;\\" width=\\"102\\" x=\\"6\\" y=\\"172\\"/><ellipse cx=\\"21\\" cy=\\"188\\" fill=\\"#B4A7E5\\" rx=\\"11\\" ry=\\"11\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><path d=\\"M16.9277,183.7651 L16.9277,181.6069 L24.3071,181.6069 L24.3071,183.7651 L21.8418,183.7651 L21.8418,191.8418 L24.3071,191.8418 L24.3071,194 L16.9277,194 L16.9277,191.8418 L19.3931,191.8418 L19.3931,183.7651 Z \\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"12\\" font-style=\\"italic\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"70\\" x=\\"35\\" y=\\"192.5352\\">order-items</text><line style=\\"stroke: #A80036; stroke-width: 1.5;\\" x1=\\"7\\" x2=\\"107\\" y1=\\"204\\" y2=\\"204\\"/><line style=\\"stroke: #A80036; stroke-width: 1.5;\\" x1=\\"7\\" x2=\\"107\\" y1=\\"212\\" y2=\\"212\\"/><!--class customer--><rect fill=\\"#FEFECE\\" filter=\\"url(#fxhzbrp8gaet7)\\" height=\\"48\\" id=\\"customer\\" style=\\"stroke: #A80036; stroke-width: 1.5;\\" width=\\"85\\" x=\\"219.5\\" y=\\"172\\"/><ellipse cx=\\"234.5\\" cy=\\"188\\" fill=\\"#ADD1B2\\" rx=\\"11\\" ry=\\"11\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><path d=\\"M237.4731,193.6431 Q236.8921,193.9419 236.2529,194.0913 Q235.6138,194.2407 234.9082,194.2407 Q232.4014,194.2407 231.0815,192.5889 Q229.7617,190.937 229.7617,187.8159 Q229.7617,184.6865 231.0815,183.0347 Q232.4014,181.3828 234.9082,181.3828 Q235.6138,181.3828 236.2612,181.5322 Q236.9087,181.6816 237.4731,181.9805 L237.4731,184.7031 Q236.8423,184.1221 236.2488,183.8523 Q235.6553,183.5825 235.0244,183.5825 Q233.6797,183.5825 232.9949,184.6492 Q232.3101,185.7158 232.3101,187.8159 Q232.3101,189.9077 232.9949,190.9744 Q233.6797,192.041 235.0244,192.041 Q235.6553,192.041 236.2488,191.7712 Q236.8423,191.5015 237.4731,190.9204 Z \\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"12\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"53\\" x=\\"248.5\\" y=\\"192.5352\\">customer</text><line style=\\"stroke: #A80036; stroke-width: 1.5;\\" x1=\\"220.5\\" x2=\\"303.5\\" y1=\\"204\\" y2=\\"204\\"/><line style=\\"stroke: #A80036; stroke-width: 1.5;\\" x1=\\"220.5\\" x2=\\"303.5\\" y1=\\"212\\" y2=\\"212\\"/><!--link order to order-items--><path d=\\"M104.9791,69.196 C76.005,80.041 45.0162,97.467 28,125 C19.8606,138.1699 25.335,154.0259 33.6257,167.3144 \\" fill=\\"none\\" id=\\"order-order-items\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><polygon fill=\\"#A80036\\" points=\\"36.5298,171.7149,34.9113,161.9999,33.7758,167.5417,28.2341,166.4063,36.5298,171.7149\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"13\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"62\\" x=\\"29\\" y=\\"138.5684\\">collection</text><!--link order to order-items--><path d=\\"M138.196,95.255 C119.051,118.773 95.8698,147.2505 79.1852,167.7466 \\" fill=\\"none\\" id=\\"order-order-items\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><polygon fill=\\"#A80036\\" points=\\"75.9381,171.7356,84.7221,167.2812,79.0947,167.858,78.5179,162.2306,75.9381,171.7356\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"13\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"35\\" x=\\"113\\" y=\\"138.5684\\">items</text><!--link order to customer--><path d=\\"M186.185,95.224 C191.954,110.644 199.582,127.7474 209,142 C215.103,151.2366 223.004,160.1877 230.833,168.0227 \\" fill=\\"none\\" id=\\"order-customer\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><polygon fill=\\"#A80036\\" points=\\"234.662,171.7758,231.0343,162.6194,231.0911,168.2759,225.4346,168.3327,234.662,171.7758\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"13\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"59\\" x=\\"210\\" y=\\"138.5684\\">customer</text><!--link order to customer--><path d=\\"M241.214,87.779 C254.409,97.924 266.461,110.309 274,125 C280.618,137.8967 278.627,153.7107 274.538,167.0428 \\" fill=\\"none\\" id=\\"order-customer\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><polygon fill=\\"#A80036\\" points=\\"272.949,171.8323,279.5797,164.5498,274.5235,167.0867,271.9867,162.0306,272.949,171.8323\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"13\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"25\\" x=\\"279\\" y=\\"138.5684\\">info</text><!--link order to order--><path d=\\"M241.345,43.843 C260.912,44.143 276,46.695 276,51.5 C276,55.854 263.608,58.359 246.721,59.014 \\" fill=\\"none\\" id=\\"order-order\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><polygon fill=\\"#A80036\\" points=\\"241.345,59.157,250.4492,62.9139,246.3432,59.0227,250.2343,54.9167,241.345,59.157\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"13\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"28\\" x=\\"282\\" y=\\"56.5684\\">next</text><!--link order to order--><path d=\\"M241.06,36.887 C279.749,34.097 316,38.968 316,51.5 C316,63.445 283.068,68.43 246.481,66.455 \\" fill=\\"none\\" id=\\"order-order\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><polygon fill=\\"#A80036\\" points=\\"241.06,66.113,249.7908,70.6708,246.0501,66.4273,250.2936,62.6866,241.06,66.113\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"13\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"27\\" x=\\"322\\" y=\\"56.5684\\">prev</text><!--
@startuml
interface \\"order\\"
\\"order\\" : orderNumber : number
\\"order\\" : itemCount : number
\\"order\\" : status : string
interface \\"order-items\\"
\\"order\\" - -> \\"order-items\\" : collection
\\"order\\" - -> \\"order-items\\" : items
\\"order\\" - -> \\"customer\\" : customer
\\"order\\" - -> \\"customer\\" : info
\\"order\\" - -> \\"order\\" : next
\\"order\\" - -> \\"order\\" : prev

@enduml

PlantUML version 1.2018.02(Fri Mar 09 12:20:44 EST 2018)
(GPL source distribution)
Java Runtime: Java(TM) SE Runtime Environment
JVM: Java HotSpot(TM) 64-Bit Server VM
Java Version: 1.8.0_131-b11
Operating System: Mac OS X
OS Version: 10.13.4
Default Encoding: UTF-8
Language: en
Country: US
--></g></svg>",
      "properties": Array [
        Object {
          "description": "Order number used to cross reference with database and accounting",
          "id": "orderNumber",
          "required": true,
          "type": "number",
        },
        Object {
          "description": undefined,
          "id": "itemCount",
          "required": true,
          "type": "number",
        },
        Object {
          "description": undefined,
          "id": "status",
          "required": true,
          "type": "string",
        },
      ],
      "resource": "order",
      "schema": Object {
        "additionalProperties": false,
        "properties": Object {
          "itemCount": Object {
            "type": "number",
          },
          "orderNumber": Object {
            "description": "Order number used to cross reference with database and accounting",
            "type": "number",
          },
          "status": Object {
            "enum": Array [
              "submitted",
              "pending",
              "completed",
            ],
            "type": "string",
          },
        },
        "required": Array [
          "orderNumber",
          "itemCount",
          "status",
        ],
        "type": "object",
      },
      "title": "Online Order",
      "typedef": "export interface Order {
  /**
   * Order number used to cross reference with database and accounting
   */
  orderNumber: number;
  itemCount: number;
  status: \\"submitted\\" | \\"pending\\" | \\"completed\\";
}
",
    },
    Object {
      "description": undefined,
      "diagram": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"no\\"?><svg xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" contentScriptType=\\"application/ecmascript\\" contentStyleType=\\"text/css\\" height=\\"192px\\" preserveAspectRatio=\\"none\\" style=\\"width:125px;height:192px;\\" version=\\"1.1\\" viewBox=\\"0 0 125 192\\" width=\\"125px\\" zoomAndPan=\\"magnify\\"><defs><filter height=\\"300%\\" id=\\"fjyfbfsvpdtwc\\" width=\\"300%\\" x=\\"-1\\" y=\\"-1\\"><feGaussianBlur result=\\"blurOut\\" stdDeviation=\\"2.0\\"/><feColorMatrix in=\\"blurOut\\" result=\\"blurOut2\\" type=\\"matrix\\" values=\\"0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 .4 0\\"/><feOffset dx=\\"4.0\\" dy=\\"4.0\\" in=\\"blurOut2\\" result=\\"blurOut3\\"/><feBlend in=\\"SourceGraphic\\" in2=\\"blurOut3\\" mode=\\"normal\\"/></filter></defs><g><!--class order-items--><rect fill=\\"#FEFECE\\" filter=\\"url(#fjyfbfsvpdtwc)\\" height=\\"48\\" id=\\"order-items\\" style=\\"stroke: #A80036; stroke-width: 1.5;\\" width=\\"102\\" x=\\"12.3519\\" y=\\"133\\"/><ellipse cx=\\"27.3519\\" cy=\\"149\\" fill=\\"#B4A7E5\\" rx=\\"11\\" ry=\\"11\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><path d=\\"M23.2797,144.7651 L23.2797,142.6069 L30.6591,142.6069 L30.6591,144.7651 L28.1937,144.7651 L28.1937,152.8418 L30.6591,152.8418 L30.6591,155 L23.2797,155 L23.2797,152.8418 L25.745,152.8418 L25.745,144.7651 Z \\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"12\\" font-style=\\"italic\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"70\\" x=\\"41.3519\\" y=\\"153.5352\\">order-items</text><line style=\\"stroke: #A80036; stroke-width: 1.5;\\" x1=\\"13.3519\\" x2=\\"113.3519\\" y1=\\"165\\" y2=\\"165\\"/><line style=\\"stroke: #A80036; stroke-width: 1.5;\\" x1=\\"13.3519\\" x2=\\"113.3519\\" y1=\\"173\\" y2=\\"173\\"/><!--class order--><rect fill=\\"#FEFECE\\" filter=\\"url(#fjyfbfsvpdtwc)\\" height=\\"48\\" id=\\"order\\" style=\\"stroke: #A80036; stroke-width: 1.5;\\" width=\\"64\\" x=\\"31.3519\\" y=\\"8\\"/><ellipse cx=\\"46.3519\\" cy=\\"24\\" fill=\\"#ADD1B2\\" rx=\\"11\\" ry=\\"11\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><path d=\\"M49.325,29.6431 Q48.744,29.9419 48.1048,30.0913 Q47.4657,30.2407 46.7601,30.2407 Q44.2533,30.2407 42.9334,28.5889 Q41.6136,26.937 41.6136,23.8159 Q41.6136,20.6865 42.9334,19.0347 Q44.2533,17.3828 46.7601,17.3828 Q47.4657,17.3828 48.1131,17.5322 Q48.7606,17.6816 49.325,17.9805 L49.325,20.7031 Q48.6942,20.1221 48.1007,19.8523 Q47.5072,19.5825 46.8763,19.5825 Q45.5316,19.5825 44.8468,20.6492 Q44.162,21.7158 44.162,23.8159 Q44.162,25.9077 44.8468,26.9744 Q45.5316,28.041 46.8763,28.041 Q47.5072,28.041 48.1007,27.7712 Q48.6942,27.5015 49.325,26.9204 Z \\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"12\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"32\\" x=\\"60.3519\\" y=\\"28.5352\\">order</text><line style=\\"stroke: #A80036; stroke-width: 1.5;\\" x1=\\"32.3519\\" x2=\\"94.3519\\" y1=\\"40\\" y2=\\"40\\"/><line style=\\"stroke: #A80036; stroke-width: 1.5;\\" x1=\\"32.3519\\" x2=\\"94.3519\\" y1=\\"48\\" y2=\\"48\\"/><!--link order to order-items--><path d=\\"M31.3007,55.765 C22.125,64.119 13.2803,74.407 8.3519,86 C1.6092,101.8614 11.1774,117.1869 24.1255,129.4336 \\" fill=\\"none\\" id=\\"order-order-items\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><polygon fill=\\"#A80036\\" points=\\"27.8564,132.804,23.8594,123.8026,24.1462,129.4522,18.4966,129.739,27.8564,132.804\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"13\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"62\\" x=\\"9.3519\\" y=\\"99.5684\\">collection</text><!--link order to order-items--><path d=\\"M70.3393,56.258 C72.757,65.433 75.1584,76.124 76.3519,86 C78.0277,99.8666 75.6912,115.1303 72.5579,127.856 \\" fill=\\"none\\" id=\\"order-order-items\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><polygon fill=\\"#A80036\\" points=\\"71.2756,132.7804,77.4145,125.0789,72.5356,127.9418,69.6727,123.0628,71.2756,132.7804\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"13\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"35\\" x=\\"77.3519\\" y=\\"99.5684\\">items</text><!--
@startuml
interface \\"order-items\\"
\\"order\\" - -> \\"order-items\\" : collection
\\"order\\" - -> \\"order-items\\" : items

@enduml

PlantUML version 1.2018.02(Fri Mar 09 12:20:44 EST 2018)
(GPL source distribution)
Java Runtime: Java(TM) SE Runtime Environment
JVM: Java HotSpot(TM) 64-Bit Server VM
Java Version: 1.8.0_131-b11
Operating System: Mac OS X
OS Version: 10.13.4
Default Encoding: UTF-8
Language: en
Country: US
--></g></svg>",
      "properties": Array [],
      "resource": "order-items",
      "schema": Object {},
      "title": undefined,
      "typedef": "export interface OrderItems {
  [k: string]: any;
}
",
    },
    Object {
      "description": undefined,
      "diagram": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"no\\"?><svg xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" contentScriptType=\\"application/ecmascript\\" contentStyleType=\\"text/css\\" height=\\"217px\\" preserveAspectRatio=\\"none\\" style=\\"width:131px;height:217px;\\" version=\\"1.1\\" viewBox=\\"0 0 131 217\\" width=\\"131px\\" zoomAndPan=\\"magnify\\"><defs><filter height=\\"300%\\" id=\\"f5wyowzmei2d9\\" width=\\"300%\\" x=\\"-1\\" y=\\"-1\\"><feGaussianBlur result=\\"blurOut\\" stdDeviation=\\"2.0\\"/><feColorMatrix in=\\"blurOut\\" result=\\"blurOut2\\" type=\\"matrix\\" values=\\"0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 .4 0\\"/><feOffset dx=\\"4.0\\" dy=\\"4.0\\" in=\\"blurOut2\\" result=\\"blurOut3\\"/><feBlend in=\\"SourceGraphic\\" in2=\\"blurOut3\\" mode=\\"normal\\"/></filter></defs><g><!--class customer--><rect fill=\\"#FEFECE\\" filter=\\"url(#f5wyowzmei2d9)\\" height=\\"73.9102\\" id=\\"customer\\" style=\\"stroke: #A80036; stroke-width: 1.5;\\" width=\\"114\\" x=\\"6\\" y=\\"133\\"/><ellipse cx=\\"34.05\\" cy=\\"149\\" fill=\\"#B4A7E5\\" rx=\\"11\\" ry=\\"11\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><path d=\\"M29.9777,144.7651 L29.9777,142.6069 L37.3571,142.6069 L37.3571,144.7651 L34.8918,144.7651 L34.8918,152.8418 L37.3571,152.8418 L37.3571,155 L29.9777,155 L29.9777,152.8418 L32.4431,152.8418 L32.4431,144.7651 Z \\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"12\\" font-style=\\"italic\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"53\\" x=\\"50.95\\" y=\\"153.5352\\">customer</text><line style=\\"stroke: #A80036; stroke-width: 1.5;\\" x1=\\"7\\" x2=\\"119\\" y1=\\"165\\" y2=\\"165\\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"11\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"102\\" x=\\"12\\" y=\\"179.6348\\">customerId : string</text><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"11\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"70\\" x=\\"12\\" y=\\"192.5898\\">name : string</text><line style=\\"stroke: #A80036; stroke-width: 1.5;\\" x1=\\"7\\" x2=\\"119\\" y1=\\"198.9102\\" y2=\\"198.9102\\"/><!--class order--><rect fill=\\"#FEFECE\\" filter=\\"url(#f5wyowzmei2d9)\\" height=\\"48\\" id=\\"order\\" style=\\"stroke: #A80036; stroke-width: 1.5;\\" width=\\"64\\" x=\\"31\\" y=\\"8\\"/><ellipse cx=\\"46\\" cy=\\"24\\" fill=\\"#ADD1B2\\" rx=\\"11\\" ry=\\"11\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><path d=\\"M48.9731,29.6431 Q48.3921,29.9419 47.7529,30.0913 Q47.1138,30.2407 46.4082,30.2407 Q43.9014,30.2407 42.5815,28.5889 Q41.2617,26.937 41.2617,23.8159 Q41.2617,20.6865 42.5815,19.0347 Q43.9014,17.3828 46.4082,17.3828 Q47.1138,17.3828 47.7612,17.5322 Q48.4087,17.6816 48.9731,17.9805 L48.9731,20.7031 Q48.3423,20.1221 47.7488,19.8523 Q47.1553,19.5825 46.5244,19.5825 Q45.1797,19.5825 44.4949,20.6492 Q43.8101,21.7158 43.8101,23.8159 Q43.8101,25.9077 44.4949,26.9744 Q45.1797,28.041 46.5244,28.041 Q47.1553,28.041 47.7488,27.7712 Q48.3423,27.5015 48.9731,26.9204 Z \\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"12\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"32\\" x=\\"60\\" y=\\"28.5352\\">order</text><line style=\\"stroke: #A80036; stroke-width: 1.5;\\" x1=\\"32\\" x2=\\"94\\" y1=\\"40\\" y2=\\"40\\"/><line style=\\"stroke: #A80036; stroke-width: 1.5;\\" x1=\\"32\\" x2=\\"94\\" y1=\\"48\\" y2=\\"48\\"/><!--link order to customer--><path d=\\"M30.9478,54.804 C16.3654,67.642 3.6258,84.882 10,103 C13.1686,112.0064 18.1425,120.6922 23.8012,128.6402 \\" fill=\\"none\\" id=\\"order-customer\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><polygon fill=\\"#A80036\\" points=\\"27.0083,132.9771,24.8734,123.3624,24.0355,128.9569,18.441,128.1189,27.0083,132.9771\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"13\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"59\\" x=\\"11\\" y=\\"99.5684\\">customer</text><!--link order to customer--><path d=\\"M69.3873,56.028 C71.6406,65.267 73.889,76.062 75,86 C76.5137,99.54 75.3173,114.2718 73.1899,127.5146 \\" fill=\\"none\\" id=\\"order-customer\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><polygon fill=\\"#A80036\\" points=\\"72.2977,132.7034,77.7652,124.5115,73.1451,127.7757,69.8809,123.1557,72.2977,132.7034\\" style=\\"stroke: #A80036; stroke-width: 1.0;\\"/><text fill=\\"#000000\\" font-family=\\"sans-serif\\" font-size=\\"13\\" lengthAdjust=\\"spacingAndGlyphs\\" textLength=\\"25\\" x=\\"76\\" y=\\"99.5684\\">info</text><!--
@startuml
interface \\"customer\\"
\\"customer\\" : customerId : string
\\"customer\\" : name : string
\\"order\\" - -> \\"customer\\" : customer
\\"order\\" - -> \\"customer\\" : info

@enduml

PlantUML version 1.2018.02(Fri Mar 09 12:20:44 EST 2018)
(GPL source distribution)
Java Runtime: Java(TM) SE Runtime Environment
JVM: Java HotSpot(TM) 64-Bit Server VM
Java Version: 1.8.0_131-b11
Operating System: Mac OS X
OS Version: 10.13.4
Default Encoding: UTF-8
Language: en
Country: US
--></g></svg>",
      "properties": Array [
        Object {
          "description": undefined,
          "id": "customerId",
          "required": true,
          "type": "string",
        },
        Object {
          "description": undefined,
          "id": "name",
          "required": true,
          "type": "string",
        },
      ],
      "resource": "customer",
      "schema": Object {
        "additionalProperties": false,
        "properties": Object {
          "customerId": Object {
            "type": "string",
          },
          "name": Object {
            "type": "string",
          },
        },
        "required": Array [
          "customerId",
          "name",
        ],
        "type": "object",
      },
      "title": undefined,
      "typedef": "export interface Customer {
  customerId: string;
  name: string;
}
",
    },
  ],
}
`;
